### PR TITLE
Feature/a1 mtdc@172 registry stub

### PR DESCRIPTION
**Pull request by algattik**
Use stub registry to resolve connector addresses. Created class structure to plug in recursive retrieval in later PR.

NB: StubRegistryClient and its JSON config mapping classes are copied from prs-composite-client.

Also factored out code to assemble parts tree to better support testing and simplify the next PR for recursive retrieval.

Deploy run: https://github.com/catenax/tractusx/actions/runs/1505010862

### CODE REVIEW

File: coreservices/partsrelationshipservice/connector/edc-recursive-job/pom.xml [hunk 1]
```diff
@@ -29,6 +29,12 @@
             <version>3.0.1</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/edc-recursive-job/pom.xml:32]: Is this dependency required?

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/extension/ApiEndpointExtension.java [hunk 2]
```diff
@@ -57,9 +101,25 @@
     @Override
     public void initialize(final ServiceExtensionContext context) {
         final var storageAccountName = ofNullable(context.getSetting(EDC_STORAGE_ACCOUNT_NAME, null))
-                .orElseThrow(() -> new EdcException("Missing mandatory property " + EDC_STORAGE_ACCOUNT_NAME));
+                .orElseThrow(() -> fatal(context, "Missing mandatory property " + EDC_STORAGE_ACCOUNT_NAME, null));
 
         final var monitor = context.getMonitor();
+        final var jsonUtil = new JsonUtil(monitor);
+
+        final var partitionsConfig = readJson(
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/extension/ApiEndpointExtension.java:109]: Consider creating a `StubRegistryFactory` containing all details of reading the json files and creating the client.

File: coreservices/partsrelationshipservice/connector/edc-recursive-job/pom.xml (hunk 1 above)
Comment (algattik) [reply coreservices/partsrelationshipservice/connector/edc-recursive-job/pom.xml:32]: Yes we're using it as EDC code does (NotNull annotations). The implicit transitive dependency is made explicit per best practice.

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/registry/StubRegistryClient.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        return getAsOptional(oneIdToUrlMappings, request.getOneIDManufacturer());
+    }
+
+    /**
+     * Helper class to manage pairs of items.
+     *
+     * @param <K> type of the first item in the pair.
+     * @param <V> type of the second item in the pair.
+     */
+    private static final class ItemPair<K, V> extends AbstractMap.SimpleEntry<K, V> {
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/registry/StubRegistryClient.java:80]: We could think of introducing https://www.javatuples.org/. WDYT?

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/DataRequestGenerator.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Generates EDC {@link DataRequest}s populated for calling Providers to invoke the PRS API
+ * to retrieve partial parts trees.
+ */
+@RequiredArgsConstructor
+@SuppressWarnings("PMD.GuardLogStatement") // Monitor doesn't offer guard statements
+public class DataRequestGenerator {
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/DataRequestGenerator.java:36]: nit: wouldn't `factory` be the name of this pattern?

File: coreservices/partsrelationshipservice/connector/edc-recursive-job/pom.xml (hunk 1 above)
Comment (cpeeyush) [reply coreservices/partsrelationshipservice/connector/edc-recursive-job/pom.xml:32]: In our code we use @Nullable from this package. So good to have it explicitly mentioned.

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java [hunk 5]
```diff
@@ -62,23 +40,23 @@
...
-    /**
-     * Blob store API.
-     */
-    private final BlobStoreApi blobStoreApi;
     /**
      * Json Converter.
      */
     private final JsonUtil jsonUtil;
+    /**
+     * XXX
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:48]: :-)

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreesAssembler.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.stream.Stream;
+
+import static java.lang.String.format;
+
+/**
+ * Assembles multiple partial parts trees into one overall parts tree.
+ */
+@RequiredArgsConstructor
+@SuppressWarnings("PMD.GuardLogStatement") // Monitor doesn't offer guard statements
+public class PartsTreesAssembler {
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreesAssembler.java:31]: Is this class related to the change "Use stub registry to resolve connector addresses"? Perhaps better separate that work?

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/registry/StubRegistryClient.java (hunk 3 above)
Comment (algattik) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/registry/StubRegistryClient.java:80]: I don't think it's worth adding a new dependency for something that can be captured in 3 lines

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/test/java/net/catenax/prs/connector/consumer/service/DataRequestGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        sut = new DataRequestGenerator(monitor, configuration, new JsonUtil(monitor), registryClient);
+    }
+
+    @Test
+    void generateRequest_WhenNoRegistryMatch_ReturnsEmpty() {
+        assertThat(sut.generateRequest(fileRequest, partId)).isEmpty();
+    }
+
+    @Test
+    void generateRequest_WhenRegistryMatch_ReturnsDataRequest() throws Exception {
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/test/java/net/catenax/prs/connector/consumer/service/DataRequestGeneratorTest.java:64]: I am not sure about the value of this test... we just duplicate the impl and compare results

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/test/java/net/catenax/prs/connector/consumer/service/DataRequestGeneratorTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                .thenReturn(Optional.of(connectorAddress));
+
+        PartsTreeByObjectIdRequest expectedPrsRequest = fileRequest.getPartsTreeRequest()
+                .toBuilder()
+                .oneIDManufacturer(partId.getOneIDManufacturer())
+                .objectIDManufacturer(partId.getObjectIDManufacturer())
+                .build();
+
+        String serializedPrsRequest = MAPPER.writeValueAsString(expectedPrsRequest);
+        var expectedRequest = DataRequest.Builder.newInstance()
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/test/java/net/catenax/prs/connector/consumer/service/DataRequestGeneratorTest.java:76]: NB: this is pretty redundant with the implementation, but will make more sense in the next PR when we will have a more generic generator method for recursion

**Review by marcgs — APPROVED** [review]

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java (hunk 5 above)
Comment (cpeeyush) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:48]: nit: more information needed ?

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/extension/ApiEndpointExtension.java (hunk 2 above)
Comment (algattik) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/extension/ApiEndpointExtension.java]: Great idea, done

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/DataRequestGenerator.java (hunk 4 above)
Comment (algattik) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/DataRequestGenerator.java:36]: Good point, done

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreesAssembler.java (hunk 6 above)
Comment (algattik) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreesAssembler.java:31]: right, I also took the opportunity to refactor code that was flagged in the previous PR. Updating the PR description.

**Review by cpeeyush — APPROVED** [review]

**Pull request merged**